apps/alert_processor/lib/model/subscription.ex

```diff
@@ -434,36 +434,96 @@ defmodule AlertProcessor.Model.Subscription do
   def all_active_for_alert(alert) do
     alert
     |> get_alert_entity_lists()
-    |> subscribers_match_query()
-    |> where_subscription_not_paused()
-    |> where_not_yet_notified(alert)
-    |> Repo.all()
-    |> Repo.preload(:user)
+    |> entity_specific_queries(alert)
   end
 
-  defp subscribers_match_query(
-         route_ids: _,
-         routes: _,
-         stops: _,
-         wildcard: true
-       ),
-       do: from(s in __MODULE__)
-
-  defp subscribers_match_query(
-         route_ids: route_ids,
-         routes: routes,
-         stops: stops,
-         wildcard: false
+  @spec entity_specific_queries(Keyword.t(), Alert.t()) :: [t()]
+  defp entity_specific_queries(
+         [route_ids: _, routes: _, stops: _, wildcard: true],
+         alert
        ) do
-    # group (route_type ANY(..) OR route ANY(..) OR origin ANY(..) OR destination ANY (...)) together
+    from(s in __MODULE__)
+    |> shared_query(alert)
+  end
+
+  defp entity_specific_queries(entity_lists, alert) do
+    entity_lists
+    |> Keyword.delete(:wildcard)
+    |> Enum.map(&entity_specific_query(&1, alert))
+    |> List.flatten(admin_query(alert))
+  end
+
+  @spec admin_query(Alert.t()) :: [t()]
+  defp admin_query(alert) do
     from(
       s in __MODULE__,
-      where:
-        s.is_admin == true or s.route_type in ^route_ids or s.route in ^routes or
-          s.origin in ^stops or s.destination in ^stops
+      where: s.is_admin == true
     )
+    |> shared_query(alert)
   end
 
+  @spec entity_specific_query(tuple(), Alert.t()) :: [t()]
+  defp entity_specific_query({:route_ids, []}, _alert), do: []
+
+  defp entity_specific_query({:route_ids, route_ids}, alert) do
+    from(
+      s in __MODULE__,
+      where: s.route_type in ^route_ids
+    )
+    |> shared_query(alert)
+  end
+
+  defp entity_specific_query({:routes, []}, _alert), do: []
+
+  defp entity_specific_query({:routes, routes}, alert) do
+    from(
+      s in __MODULE__,
+      where: s.route in ^routes
+    )
+    |> shared_query(alert)
+  end
+
+  defp entity_specific_query({:stops, []}, _alert), do: []
+
+  defp entity_specific_query({:stops, stops}, alert) do
+    from(
+      s in __MODULE__,
+      where: s.origin in ^stops or s.destination in ^stops
+    )
+    |> shared_query(alert)
+  end
+
+  defp shared_query(query, alert) do
+    query
+    |> where_subscription_not_paused()
+    |> where_not_yet_notified(alert)
+    |> Repo.all()
```

Comment:
suggestion(non-blocking): you might consider splitting this part up, so that we only fetch the user once even if they're present in multiple alerts. a way to do this would be to have the "entity specific queries" and "admin_query" return a list of IDs, rather than a full subscription. then at the end, perform one query by the subscription ID, and preload the user on that single query.

Reply:
Thanks, great idea. That made me realize I could through a `uniq` constraint in there while I was at it. Since this is a non-trivial change, would you mind giving it one last quick once-over, please?

Reply:
And actually wasn't even being used! So this saved it from deletion.